R/pecs.R

```diff
@@ -0,0 +1,139 @@
+#'   If `times` is missing or given as a range, then `n` provide number of time-points to evaluate
+#'    `meas` over.
+#' @param eps (`numeric()`) \cr
+#'   Small error value to pass to [MeasureSurvIntLogloss] to prevent errors resulting from a log(0)
+#'   calculation.
+#' @param task ([TaskSurv])
+#' @param ... Additional arguments.
+#'
+#' @details If `times` and `n` are missing then `meas` is evaluated over all observed time-points
+#' from the [PredictionSurv] or [TaskSurv] object. If a range is provided for `times` without `n`,
+#' then all time-points between the range are returned.
+#'
+#' @examples
+#' library(mlr3)
+#' task = tsk("rats")
+#'
+#' # Prediction Error Curves for prediction object
+#' learn = lrn("surv.coxph")
+#' p = learn$train(task)$predict(task)
+#' pecs(p)
+#' pecs(p, meas = "logloss", times = c(20, 40, 60, 80)) +
+#'   ggplot2::geom_point() +
+#'   ggplot2::ggtitle("Logloss Prediction Error Curve for Cox PH")
+#'
+#' # Prediction Error Curves for fitted learners
+#' learns = lrns(c("surv.kaplan", "surv.coxph", "surv.ranger"))
+#' lapply(learns, function(x) x$train(task))
+#' pecs(learns, task = task, meas = "logloss", times = c(20, 90), n = 10)
+#'
+#'
+#' @export
+pecs = function(x, meas = c("graf","logloss"), times, n, eps = 1e-15, ...){
+  if(!missing(times)) assertNumeric(times, min.len = 1)
+  if(!missing(n)) assertIntegerish(n, len = 1)
+  assertNumeric(eps, lower = -1, upper = 1)
+
+  UseMethod("pecs", x)
+}
+
+#' @rdname pecs
+#' @export
+pecs.list = function(x, meas = c("graf","logloss"), times, n, eps = 1e-15, task,...){
+  meas = match.arg(meas)
+
+  assert(all(sapply(x, function(y) !is.null(y$model))), "x must be a list of trained survival learners")
+  assertClass(task, "TaskSurv")
+
+  p = lapply(x, function(y) y$predict(task))
+  true_times = sort(unique(task$truth()[,1]))
+
+  times = .pec_times(true_times = true_times, times = times, n = n)
+
+  if(meas == "logloss"){
+    scores = lapply(p, function(y){
+      integrated_score(score = weighted_logloss(truth = task$truth(),
+                                                distribution = y$distr,
+                                                times = times,
+                                                eps = eps),
+                       integrated = FALSE)
+    })
+  } else {
+    scores = lapply(p, function(y){
+      integrated_score(score = weighted_graf(truth = task$truth(),
+                                             distribution = y$distr,
+                                             times = times),
+                       integrated = FALSE)
+    })
+  }
+
+  times = as.numeric(names(scores[[1]]))
+  scores = round(rbindlist(list(scores)), 4)
+  colnames(scores) = sapply(x, function(y) gsub("surv.", "", y$id, fixed = TRUE))
+  scores$time = times
+  scores = melt(scores, "time", value.name = meas, variable.name = "learner")
+
+  ggplot2::ggplot(data = scores, ggplot2::aes_string(x = "time", color = "learner", y = meas)) +
```

Comment:
Refactor? separate function that creates the data and is called within the autoplot function.

Reply:
maybe fix y-scale to c(0, .33)? then all plots are comparable with each other.

Reply:
> Refactor? separate function that creates the data and is called within the autoplot function.

Not entirely sure what you're suggesting here sorry

> maybe fix y-scale to c(0, .33)? then all plots are comparable with each other.

This is problematic as it stops users from manually later changing this, also I can imagine different people having different opinions about limits

Reply:
> > Refactor? separate function that creates the data and is called within the autoplot function.
> 
> Not entirely sure what you're suggesting here sorry

I think it would be better to have one separate function that produces the data set that is then used for plotting outside of pecs and then `pecs` calls this function and plots the data.

Reply:
In theory I agree but for the sake of just two functions not sure what the added benefit is as both have different `ggplot` calls anyway. But happy to hear suggestions and how you think it could be made more efficient

Reply:
its not so much about efficiency. I personally just like to have the option to  call a function that returns the data used for plotting, so I can do my own plots. But its nice to have convenience plotting functions for quick overview. But I guess its not too important for now. We can still refactor later, as long as the interface of `pecs` doesn't change.

Reply:
So what I've done in my plotting function in `distr6` is to have an additional argument `plot`, which can be set to `FALSE` and in this case it returns the data structure internally, how about that here?